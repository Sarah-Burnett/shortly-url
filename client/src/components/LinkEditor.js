import React, { useState, useContext } from 'react';
import { Button, PrimaryButton, LinkModal, Error } from '../Styles';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faExclamationTriangle } from '@fortawesome/free-solid-svg-icons';
import { LinkUpdateContext } from '../context/LinkUpdateContext';
import { LinksContext } from '../context/LinksContext';
import axios from 'axios';

export const LinkEditor = () => {
    const [ links, setLinks ] = useContext(LinksContext);
    const [ linkUpdate, setLinkUpdate ] = useContext(LinkUpdateContext);
    const [ newLink, setNewLink ] = useState(linkUpdate.link.short);
    const [ error, setError ] = useState();
    const handleChange = (event) => {
        setNewLink(event.target.value);
        setError(null);
    };
    const updateLink = (event) => {
        const {link} = linkUpdate;
        if (newLink === link._id) setError('Url same as existing');
        link.short = newLink;
        axios.put(`/api/links/${link._id}`, link)
        .then( res => {
                setLinkUpdate(''); 
                setLinks( prev => prev.map( prevLink => {
                    return prevLink._id === link._id ? {...prevLink, short: link.short} : {...prevLink}
                }))
        })
        .catch( err => setError('Update failed, please try again'));
    }
    return (
            <LinkModal>
                <h3>Edit Shortly Link</h3>
                <p>{linkUpdate.link.full}</p>
                <div style={{display: "flex"}}>
                    <p>https://shortly/</p>
                    <input onChange={handleChange} value={newLink}/>
                </div>
                { error && <Error><FontAwesomeIcon icon={faExclamationTriangle}/> {error}</Error> }
                <PrimaryButton id="update" type="button" onClick={() => updateLink()}>Update</PrimaryButton>
                <Button type="button" onClick={() => setLinkUpdate('')}>Cancel</Button>
            </LinkModal> 
    )
}
