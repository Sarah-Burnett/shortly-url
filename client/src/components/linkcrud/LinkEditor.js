import React, { useState } from 'react';
import { PrimaryButton, LinkModal } from '../Styles';
import axios from 'axios';
import { shortlyUrl } from '../../utilities/urls';
import Error from '../Error';
import CloseButton from './CloseButton';
import { useProcessLink } from '../../hooks/useProcessLink';
import { useLinks } from '../../hooks/useLinks';
import { useFlash } from '../../hooks/useFlash';

const LinkEditor = () => {
    const { updateLinkInLinks } = useLinks();
    const { processLink: {link, link: {_id, full, short}}, setProcessNull } = useProcessLink();
    const [ newShortLink, setNewShortLink ] = useState(short);
    const { showFlash } = useFlash();
    const [ error, setError ] = useState();
    const handleChange = (event) => {
        setError(null);
        setNewShortLink(event.target.value);
    };
    const updateLink = (event) => {
        event.preventDefault();
        if (newShortLink === short) return setError('Url same as existing');
        const newLink = {...link, short: newShortLink};
        axios.put(`/api/links/${_id}`, newLink)
        .then( () => {
            updateLinkInLinks(newLink);
            showFlash('Link updated');
            setProcessNull(); 
        })            
        .catch( () => setError('Update failed, please try again'));
    }
    return (
            <LinkModal onSubmit={(event) => updateLink(event)}>
                    <h3>Edit Shortly Link</h3>
                    <p>{full}</p>
                    <div style={{display: "flex"}}>
                        <p>{shortlyUrl}</p>
                        <input 
                            onChange={handleChange} 
                            value={newShortLink || ''} 
                            required 
                            pattern="^[a-z0-9_-]{3,16}$" 
                            title="Letters and numbers between 3-16 characters"
                            data-testid="input"
                        />
                    </div>
                    { error && <Error error={error}/> }
                    <PrimaryButton id="update" data-testid="button">Update</PrimaryButton>
                    <CloseButton/>
            </LinkModal> 
    )
};

export default LinkEditor;
