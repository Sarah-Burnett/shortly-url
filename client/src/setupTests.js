import React from 'react';
import ReactDOM from 'react-dom';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup, fireEvent, screen, waitForElement } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import renderer from 'react-test-renderer';
import { UserProvider } from './context/UserContext';
import { LinksProvider } from './context/LinksContext';
import { ActiveLinkProvider } from './context/ActiveLinkContext';
import { ProcessLinkProvider } from './context/ProcessLinkContext';
import { FlashProvider } from './context/FlashContext';
import App from './App';

export const ComponentWrapper = (Component) => {
    return (
        <UserProvider>
            <LinksProvider>
                <ActiveLinkProvider>
                    <ProcessLinkProvider>
                        <FlashProvider>
                            <Component/>
                        </FlashProvider>
                    </ProcessLinkProvider>
                </ActiveLinkProvider>
            </LinksProvider>
        </UserProvider>
    )
}

export const rendersWithoutCrashing = (Component) => {
    const div = document.createElement('div');
    ReactDOM.render(<Component/>, div);
}

export const matchesSnapshot = (Component) => {
    const tree = renderer.create(<Component/>).toJSON();
    expect(tree).toMatchSnapshot();
}

export const sampleLinks = [
    {
        "date":"2020-08-14T17:29:53.041Z",
        "click":[
            {"date":1597426302926},
            {"date":1597426306071},
            {"date":1597426310231},
            {"date":1597426314763},
            {"date":1597428206995},
            {"date":1597428210093},
            {"date":1597428215131},
            {"date":1597428218755}
        ],
        "_id":"K1pjMCK-F",
        "full":"http://reactcommunity.org/react-transition-group/css-transition",
        "userId":"5ebe7d235e4a501ee8699058",
        "short":"octopus","__v":8
    },
    {
        "date":"2020-08-14T17:33:55.748Z",
        "click":[],
        "_id":"lsncCe5RV",
        "full":"https://ba-grant.netlify.app/",
        "userId":"5ebe7d235e4a501ee8699058",
        "short":"bagrant","__v":0
    }
];
