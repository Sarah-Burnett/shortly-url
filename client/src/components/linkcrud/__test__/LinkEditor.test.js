import { rendersWithoutCrashing, matchesSnapshot, ComponentWrapper } from '../../../setupTests';
import LinkEditor from '../LinkEditor';
import { render, fireEvent, screen } from '@testing-library/react';

const LinkEditorWrapper = () => ComponentWrapper(LinkEditor)

it('renders link editor without crashing', () => rendersWithoutCrashing(LinkEditorWrapper));

it('link editor matches snapshot', () => matchesSnapshot(LinkEditorWrapper));

it('valid new short url does not show error', async () => {
    render(LinkEditorWrapper());
    await fireEvent.change(screen.getByTestId('input'), {
        target: {
            value: "shortLink"
        }
    })
    await fireEvent.click(screen.getByTestId('button'))
    expect(screen.queryByTestId('error')).toBe(null);
})

it('invalid new short url shows error', async () => {
    render(LinkEditorWrapper());
    await fireEvent.change(screen.getByTestId('input'), {
        target: {
            value: ""
        }
    })
    await fireEvent.click(screen.getByTestId('button'))
    expect(screen.getByTestId('error')).toHaveTextContent('Please enter a new short link');
})
