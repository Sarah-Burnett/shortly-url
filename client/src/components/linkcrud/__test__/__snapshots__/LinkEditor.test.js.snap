// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`link editor matches snapshot 1`] = `
<form
  className="sc-fzozJi eAysAZ"
  onSubmit={[Function]}
>
  <h3>
    Edit Shortly Link
  </h3>
  <p />
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <p>
      https://short.ly/
    </p>
    <input
      data-testid="input"
      onChange={[Function]}
      pattern="^[a-z0-9_-]{3,16}$"
      required={true}
      title="Letters and numbers between 3-16 characters"
    />
  </div>
  <button
    className="sc-AxirZ fTDgvF"
    data-testid="button"
    id="update"
  >
    Update
  </button>
  <button
    className="sc-AxjAm iwCWZU"
    onClick={[Function]}
    type="button"
  >
    Close
  </button>
</form>
`;
